### ISSUE TITLE
Snowflake: Interaction between L025 and L028 corrupts SQL

### ISSUE THREAD

**Issue by WittierDinosaur** [sqlfluff/sqlfluff#3178]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqfluff fix 
```
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, LATERAL FLATTEN(input => rec:result) x
;
```
Produces:
```
L:   3 | P:   5 | L028 | Unqualified reference 'value' found in single table select.
L:   4 | P:  54 | L025 | Alias 'x' is never used in SELECT statement.

SELECT
  r.rec:foo::string AS foo
  , r.value:bar::string AS bar
FROM foo.bar r, LATERAL FLATTEN(input => rec:result)
;
```
The issue here is it doesn't recognise the `LATERAL FLATTEN` as a table entity. Replacing the `LATERAL FLATTEN` with `bar` gives

```
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, bar x
;

L:   4 | P:  21 | L025 | Alias 'x' is never used in SELECT statement.
```

### Expected Behaviour

LATERAL FLATTEN should be treated like a table

### Observed Behaviour

-

### How to reproduce

-

### Dialect

Snowflake

### Version

0.13.0

### Configuration

Unaffected by config

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
@WittierDinosaur I'm a little confused so think might need some clarifications here.

BigQuery has the `value_table_functions` for things like `UNNEST` and looks like that be similar.

But still confused as `x` is not used. Should it be `x.value.:bar::string`?

> The issue here is it doesn't recognise the LATERAL FLATTEN as a table entity. Replacing the LATERAL FLATTEN with bar gives

You didn't replace that with `bar`?

Are your examples correct?

**Comment by WittierDinosaur** [comment]
oops - corrected. Pasted the wrong SQL string into the second example

**Comment by tunetheweb** [comment]
Something still not right. Both the first and the second SQL produce the exact same errors on `lint`. And it's L027 not L028 it's complaining about.

**Comment by WittierDinosaur** [comment]
is that true on sqlfluff fix for the two? I just retested and I can verify both of them

**Comment by WittierDinosaur** [comment]
i made a correction twice, maybe you grabbed the sql between the two?

**Comment by tunetheweb** [comment]
First SQL:

```
% sqlfluff lint --dialect snowflake test.sql                                                                              ((HEAD detached at 0.12.0))
== [test.sql] FAIL                                                                                                                                                                          
L:   2 | P:   3 | L003 | Expected 1 indentations, found 0 [compared to line 01]
L:   3 | P:   3 | L003 | Expected 1 indentations, found 0 [compared to line 01]
L:   3 | P:   3 | L019 | Found leading comma. Expected only trailing.
L:   3 | P:   5 | L027 | Unqualified reference 'value' found in select with more
                       | than one referenced table/view.
L:   4 | P:  14 | L011 | Implicit/explicit aliasing of table.
L:   4 | P:  14 | L031 | Avoid aliases in from clauses and join conditions.
L:   4 | P:  54 | L011 | Implicit/explicit aliasing of table.
L:   4 | P:  54 | L025 | Alias 'x' is never used in SELECT statement.
L:   4 | P:  55 | L052 | Statements must end with a semi-colon.
All Finished 📜 🎉!
```

Second SQL:
```
% sqlfluff lint --dialect snowflake test.sql                                                                              ((HEAD detached at 0.12.0))
== [test.sql] FAIL                                                                                                                                                                          
L:   2 | P:   3 | L003 | Expected 1 indentations, found 0 [compared to line 01]
L:   3 | P:   3 | L003 | Expected 1 indentations, found 0 [compared to line 01]
L:   3 | P:   3 | L019 | Found leading comma. Expected only trailing.
L:   3 | P:   5 | L027 | Unqualified reference 'value' found in select with more
                       | than one referenced table/view.
L:   4 | P:  14 | L011 | Implicit/explicit aliasing of table.
L:   4 | P:  14 | L031 | Avoid aliases in from clauses and join conditions.
L:   4 | P:  21 | L011 | Implicit/explicit aliasing of table.
L:   4 | P:  21 | L025 | Alias 'x' is never used in SELECT statement.
L:   4 | P:  21 | L031 | Avoid aliases in from clauses and join conditions.
L:   4 | P:  22 | L052 | Statements must end with a semi-colon.
All Finished 📜 🎉!
```

Note I have a lot more errors, because you did not provide your SQLFluff config. It would be nice if maintainers set and example here ;-)

**Comment by WittierDinosaur** [comment]
That's fair - can't see anything in the config that'd cause this 
```
[sqlfluff]
# https://docs.sqlfluff.com/en/stable/configuration.html#default-configuration
verbose = 0
nocolor = False
dialect = snowflake
templater = jinja
rules = None
exclude_rules = L011,L016,L031,L034,L035
recurse = 0
output_line_length = 120
runaway_limit = 10
ignore = None
ignore_templated_areas = True
encoding = autodetect
disable_noqa = False
sql_file_exts = .sql,.sql.j2,.dml,.ddl
fix_even_unparsable = False

[sqlfluff:indentation]
indented_joins = False
indented_ctes = False
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable, default='') %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 2
max_line_length = 120
indent_unit = space
comma_style = leading
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
# https://docs.sqlfluff.com/en/stable/rules.html
[sqlfluff:rules:L003]
hanging_indents = False

[sqlfluff:rules:L007]  # Keywords
operator_new_lines = after

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = upper

[sqlfluff:rules:L011]  # Table aliasing
aliasing = implicit

[sqlfluff:rules:L012]  # Column aliasing
aliasing = explicit

[sqlfluff:rules:L013]  # Column expressions
allow_scalar = False

[sqlfluff:rules:L014]  # Unquoted identifiers
extended_capitalisation_policy = lower

[sqlfluff:rules:L016]
ignore_comment_lines = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases

[sqlfluff:rules:L030]  # Function names
extended_capitalisation_policy = upper

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = consistent

[sqlfluff:rules:L042]
# By default, allow subqueries in FROM clauses, but not JOIN clauses.
forbid_subquery_in = join

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False

[sqlfluff:rules:L052]  # Statements must end with a semi-colon
multiline_newline = True
require_final_semicolon = True

[sqlfluff:rules:L059] # Unnecessary quoted identifier
ignore_words = date
```

**Comment by WittierDinosaur** [comment]
Just to check, you're linting this for the first?
```
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, LATERAL FLATTEN(input => rec:result) x
;
```

**Comment by WittierDinosaur** [comment]
Oh jeez - the output is different for lint and fix
```
sqlfluff lint test.sql
== [test.sql] FAIL                                                                                         
L:   3 | P:   5 | L027 | Unqualified reference 'value' found in select with more than one referenced table/view.
L:   4 | P:  54 | L025 | Alias 'x' is never used in SELECT statement.
All Finished 📜 🎉!


sqlfluff fix test.sql
==== finding fixable violations ====
== [test.sql] FAIL                                                                                         
L:   3 | P:   5 | L028 | Unqualified reference 'value' found in single table select.
L:   4 | P:  54 | L025 | Alias 'x' is never used in SELECT statement.
==== fixing violations ====
2 fixable linting violations found
```

**Comment by tunetheweb** [comment]
Well that doesn't help!

OK let's look at `lint` first.

you agree they both show the same problem?

```
(.venv) barry@Barrys-MacBook sqlfluff % cat test.sql
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, LATERAL FLATTEN(input => rec:result) x
;
(.venv) barry@Barrys-MacBook sqlfluff % sqlfluff lint --dialect snowflake test.sql                                                                              ((HEAD detached at 0.12.0))
== [test.sql] FAIL
L:   3 | P:   5 | L027 | Unqualified reference 'value' found in select with more than one referenced table/view.
L:   4 | P:  54 | L025 | Alias 'x' is never used in SELECT statement.
All Finished 📜 🎉!
(.venv) barry@Barrys-MacBook sqlfluff % cat test2.sql
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, bar x
;
(.venv) barry@Barrys-MacBook sqlfluff % sqlfluff lint --dialect snowflake test2.sql                                                                             ((HEAD detached at 0.12.0))
== [test2.sql] FAIL
L:   3 | P:   5 | L027 | Unqualified reference 'value' found in select with more than one referenced table/view.
L:   4 | P:  21 | L025 | Alias 'x' is never used in SELECT statement.
All Finished 📜 🎉!
```

And to be honest I agree with the two errors. But then maybe I'm missing something here?

**Comment by tunetheweb** [comment]
What's interesting is that L027 is not `fix` compatible, but L028 is.

So for `lint` it recognises the real problem is L027, and therefore doesn't seem to then also include this in L028.
But for `fix`, cause L027 is not in play (as it's not `fix`-compatible) it does include this in L028.

Weird!

@barrywhart any ideas on this one? It's an odd one alright!

**Comment by barrywhart** [comment]
Is this the query we're discussing?

```
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, bar x
```

That's an unusual join syntax. Do you get the same result using the equivalent `JOIN`?

**Comment by tunetheweb** [comment]
That one works and does not flag L028 on `fix`.

This SQL is basically the same but DOES flag L028 in `fix` (and weirdly not on `lint`!):

```sql
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, LATERAL FLATTEN(input => rec:result) x
;
```

**Comment by barrywhart** [comment]
No ideas offhand. Someone needs to look at the underlying rule. There's a function shared by many rules -- it returns a bunch of info, including a list of sources (aliases, tables, etc.). Good chance the issue is with that shared function and/or how the specific rule interprets the result from that function.

**Comment by barrywhart** [comment]
@WittierDinosaur: Can you provide "Expected Behavior" in the form of SQL?

I do have a partial fix for this issue, but I can't tell if it's a complete fix because I don't have this. I'll create a PR with what I have so far.

**Comment by WittierDinosaur** [comment]
Fix should leave it alone, lint should flag L027. So I would expect input to be the same as output:
```
SELECT
  r.rec:foo::string AS foo
  , value:bar::string AS bar
FROM foo.bar r, LATERAL FLATTEN(input => rec:result) x
;
```
It shouldn't remove the `x` alias, and it can't infer which alias is associated with `value`
